content/collective_intro.tex

```diff
@@ -123,6 +123,11 @@ \subsubsection*{Team-implicit collectives}
 call to \FUNC{shmem\_team\_sync} on the default team.
 
 \subsubsection*{Error codes returned from collectives}
-\CorCpp routines that return an integer error code follow the convention that \CONST{0} indicates successful local completion of the operation. This is considered a best effort of the implementation to indicate that all required local operations have been performed correctly inside the routine and the internal \openshmem state on the calling \ac{PE} is consistent with the description of the routine and its arguments upon completion. Implementations may use an integer return value from a routine to define integer error codes specific to the implementation as long as those codes are not already explicitly defined for that routine by the \openshmem specification.
 
-Collective operations involving many \acp{PE} may return values indicating success while other \acp{PE} are still executing the collective operation. Return values indicating success of a collective routine on one \ac{PE} do not indicate that all \acp{PE} involved in the collective operation will return from the routine successfully. In the case where successful local completion of a collective implies the success of some global operation, such as team creation resulting in a valid team on all involved \acp{PE}, the implementation should not return \CONST{0} from the routine if the implied or stated global guarantees of the routine are not met.
+Collective operations involving multiple \acp{PE} may return values
+indicating success while other \acp{PE} are still executing the
+collective operation. Return values indicating success or failure of a
+collective routine on one \ac{PE} may not indicate that all \acp{PE}
+involved in the collective operation will return the same value. Some
+operations, such as team creation, must return consistent return codes
+across multiple \acp{PE}.
```

Comment:
> Does "must return consistent return values" require all PEs to return _equal_ return values?

Yes, I think to say "equal" is clearer to users and implementers.

> For example, during team creation - if PEs in the active set discover they cannot successfully create a team, do those PEs need to notify the other PEs in the parent team (i.e., outside the active set) that they need to return a nonzero value?

Yes, this is required and, I would argue, the natural behavior an application developer would expect. If team creation could fail on a subset of PEs in the child team, then a portable application is forced to use the parent team to check for success across the child team. There's no guarantee that a collective on the child team would ever be successful otherwise. Executing the reduction inside the implementation is worth the cost to the end user.

> I recall discussing this at the F2F, but don't recall if we decided the cost of a reduction is worth having equal return values across PEs during team creation.

At the F2F, I recall discussing `shmem_team_split_2d` as one of the "important" cases to consider here. While we initially thought there would be a reduction involved, I think we rationalized to ourselves that if any one team on any PE in the parent team fails to create any child team, then all PEs experience the cascading failure.

Reply:
Agreed, thanks.

I may not fully understand how a "cascading failure" for 2D-splits obviates a reduction... but I'll think about it.  As of now, I think the SOS prototype might need a reduction in `split_strided`, which is called by `split_2d` at least a couple times on each PE.